app/src/main/java/com/example/drawingApp/MainActivity.kt

```diff
@@ -1,23 +1,118 @@
 package com.example.drawingApp
 
 import android.os.Bundle
+import android.os.StrictMode
+import android.view.LayoutInflater
+import android.view.View
+import android.widget.EditText
+import android.widget.TextView
+import android.widget.Toast
 import androidx.activity.viewModels
+import androidx.appcompat.app.AlertDialog
 import androidx.appcompat.app.AppCompatActivity
+import androidx.core.view.isVisible
+import androidx.core.widget.addTextChangedListener
 
 class MainActivity : AppCompatActivity() {
-    private val model: MainViewModel by viewModels()
+    private val model: DrawingViewModel by viewModels()
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
-
+        enableStrictMode()
         val colorCircleView = findViewById<SelectedColorView>(R.id.colorCircle)
         val drawingFieldView = findViewById<DrawingFieldView>(R.id.drawField)
 
+        drawingFieldView.onLineDrawing = {
+            model.activeBitmap = it
+        }
+
+        findViewById<TextView>(R.id.newLayer).setOnClickListener {
+            showLayerAlertDialog()
+        }
+
+        model.activeBitmap?.let {
+            drawingFieldView.setBitmap(it)
+            drawingFieldView.invalidate()
+        }
+
         colorCircleView.onColorChange = { stateColor ->
             model.primaryColor = stateColor
             drawingFieldView.updateColor(stateColor)
         }
 
+        if (model.isLayerCreationDialogOpen) showLayerAlertDialog()
+
+    }
+
+    fun showLayerAlertDialog() {
+        val inflater = LayoutInflater.from(this@MainActivity)
+        val view = inflater.inflate(R.layout.layer_dialog, null)
+        val layerField = view.findViewById<EditText>(R.id.newLayerField)
+        val errorText = view.findViewById<TextView>(R.id.dialogCharCountError)
+        var layerCounter = 1
+
+        //find the first available layer name not taken
+        while (model.layerNames.contains(getString(R.string.new_layer_hint, layerCounter))) {
+            layerCounter += 1
+        }
+        layerField.setText(getString(R.string.new_layer_hint, layerCounter))
+
+        //needs to be after the set text or else the textchanged listener will overwrite the previous text needed for rotations
+        layerField.addTextChangedListener {
+            model.currentLayerCreationDialogText = layerField.text.toString()
+            errorText.isVisible = it.toString().trim().length > MAX_CHAR_LEN
+            if (errorText.isVisible) {
+                layerField.setText(layerField.text.toString().dropLast(1))
```

Comment:
I don't see why you need to do this at all.... what is this solving? If you really wanted to not let them type any more you could take the first MAX_CHAR_LEN substring. If I paste 30 characters your logic fails. Or just set max length on the edittext

Reply:
if you paste 30 characters it does still work correctly, however you are correct substring would be much cleaner, if they end up trying to paste an extremely long string it would create a massive recursive call.